### ISSUE TITLE
Generated query has extra non-existent column with suffix 1

### ISSUE THREAD

**Issue by ygoe** [dotnet/efcore#26924]
After removing a column from a database table (and model class) and redirecting it as NotMapped to another related object, EF will generate invalid queries for that table. It keeps thinking that the old column is still there, but as a ghost with the name suffix "1". I have no idea why that happens and why it is possible at all. EF should rather not invent arbitrary column names but tell me what the problem is so I can fix it. Adding random column names to a query will always fail because nobody has created that column in the database.

This is the shortened code of the relevant model class:

```cs
[Table("connectors")]
public class Connector : IHasLocationId
{
	[Key]
	public int Id { get; set; }

	[Required]
	[ForeignKey(nameof(Organization))]
	public int OrganizationId { get; set; }

	public virtual Organization Organization { get; set; }
	
	// A few other properties...

	// This was a table column, now it's stored elsewhere and just redirected
	[NotMapped]
	public int? LocationId => User.LocationId;
}

public interface IHasLocationId
{
	int OrganizationId { get; }

	int? LocationId { get; }
}

// This is the code that executes the query:
var query = dbContext.Connectors
	.Where(c => c.OrganizationId == organizationId);
	.OrderBy(c => c.User.Name.ToLower());
var connectors = query.ToList();
// This generates an SQL query that includes the non-existing column "location_id1" of the "connectors" table
// which in turn generates an error from the database that doesn't know what that is.

fail: Microsoft.EntityFrameworkCore.Database.Command[20102]
	  Failed executing DbCommand (22ms) [Parameters=[@__organizationId_0='23064233'], CommandType='Text', CommandTimeout='30']
      SELECT c.id, c.description, c.location_id1, c.organization_id, c.user_id
      FROM connectors AS c
      INNER JOIN users AS u ON c.user_id = u.id
      WHERE c.organization_id = @__organizationId_0
      ORDER BY LOWER(u.name)
```

There is no symbol `LocationId1` anywhere in the entire project solution.

What's going on here and how can this be fixed? I need to complete this quickly but now I'm completely stuck with EF. The only solution I know is not using EF but that creates a lot more work than is acceptable.

### Include provider and version information

EF Core version: 3.1.15
Database provider: Postgres
Target framework: .NET Core 3.1
Operating system: Windows 10
IDE: Visual Studio 2019

**Comment by ajcvickers** [comment]
@ygoe `LocationId1` is created as a [shadow property](https://docs.microsoft.com/en-us/ef/core/modeling/shadow-properties) to act as the foreign key in a [relationship](https://docs.microsoft.com/en-us/ef/core/modeling/relationships?tabs=fluent-api%2Cfluent-api-simple-key%2Csimple-key) that exists in your model. EF would normally use `LocationId` for this, but since this property already exists in the CLR type and has been excluded with `[NotMapped]`, it means EF cannot use this name. There should be a message about this in the debug [logging](https://docs.microsoft.com/en-us/ef/core/logging-events-diagnostics/).

Typically, this happens because the relationship is incorrectly configured. However, there isn't enough information above to know for sure. If you attach a small, runnable project or post a small, runnable code listing that reproduces what you are seeing, then we can investigate more.

**Comment by ygoe** [comment]
Thank you for that hint. There was indeed another reverse collection property in the model class whose reference was removed. Removing that cleared up the error. I just forgot to remove it.

I couldn't make use of the debug log though. I think you meant to set something like this in the app config file:

```
  "Logging": {
    "LogLevel": {
      "Microsoft.EntityFrameworkCore": "Debug"
    }
  }
```

That produces so much output that the application never finishes writing the console output (Windows console windows are slow as an old landline terminal) and cannot properly be terminated anymore. I've tried to clear the output (`Console.Clear()`) and see what's printed as the broken query is executed, but there was nothing interesting but the query itself. Then I looked at what was printed up until the very first database query, when the model might be created. This was the only line I could find that relates to the problem:

```
dbug: Microsoft.EntityFrameworkCore.Model.Validation[10600]
      The property 'LocationId1' on entity type 'Connector' was created in shadow state because there are no eligible CLR members with a matching name.
```

Not a lot. And especially nothing new. So that log isn't helpful at all here. I had hoped to see an explanation why that reference should be made at all. Like a pointer to the other property (that I removed eventually) that led to this. It's in another class and is easy to forget.

Anyway, I don't understand why EF simply invents a column name here. It must be clear that this will fail and nobody would know where that strange name comes from because it wasn't specified anywhere. I think this log line should at least be in the warning level, and it should really include the reason why this reference was made at all. Maybe this issue should even throw an exception with a good explanation. I cannot imagine how the current behaviour could be of any use.

PS: To clarify things, I'm not using any kind of EF migrations. The SQL schema is managed separately, EF is only expected to generate the queries to the "external" database schema and do the O/R work.

**Comment by ajcvickers** [comment]
@ygoe See #20435, which implemented in 6.0.

**Comment by TheMrAnderson** [comment]
This is still an issue in .Net 8 and I'm left chasing my tail and fake errors are created at runtime rather than a build error stopping the app from running and giving me the possibility of figuring out the magic combination needed to get some of these more complex relationships working.
